exporter/exporterhelper/queued_retry.go

```diff
@@ -126,39 +137,142 @@ func createSampledLogger(logger *zap.Logger) *zap.Logger {
 	return logger.WithOptions(opts)
 }
 
-func newQueuedRetrySender(fullName string, qCfg QueueSettings, rCfg RetrySettings, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
+func newQueuedRetrySender(id config.ComponentID, signalName string, qCfg QueueSettings, rCfg RetrySettings, reqUnmarshaler requestUnmarshaler, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
 	retryStopCh := make(chan struct{})
 	sampledLogger := createSampledLogger(logger)
-	traceAttr := trace.StringAttribute(obsmetrics.ExporterKey, fullName)
-	return &queuedRetrySender{
-		fullName: fullName,
-		cfg:      qCfg,
-		consumerSender: &retrySender{
-			traceAttribute: traceAttr,
-			cfg:            rCfg,
-			nextSender:     nextSender,
-			stopCh:         retryStopCh,
-			logger:         sampledLogger,
-		},
-		queue:           queue.NewBoundedQueue(qCfg.QueueSize, func(item interface{}) {}),
-		retryStopCh:     retryStopCh,
-		traceAttributes: []trace.Attribute{traceAttr},
-		logger:          sampledLogger,
+	traceAttr := trace.StringAttribute(obsmetrics.ExporterKey, id.String())
+
+	qrs := &queuedRetrySender{
+		id:                 id,
+		signalName:         signalName,
+		cfg:                qCfg,
+		retryStopCh:        retryStopCh,
+		traceAttributes:    []trace.Attribute{traceAttr},
+		logger:             sampledLogger,
+		requestUnmarshaler: reqUnmarshaler,
 	}
+
+	qrs.consumerSender = &retrySender{
+		traceAttribute: traceAttr,
+		cfg:            rCfg,
+		nextSender:     nextSender,
+		stopCh:         retryStopCh,
+		logger:         sampledLogger,
+		// Following three functions are provided in such way because they actually depend on queuedRetrySender
+		onSuccess:          qrs.onSuccess,
+		onTemporaryFailure: qrs.onTemporaryFailure,
+		onPermanentFailure: qrs.onPermanentFailure,
+	}
+
+	if !qCfg.PersistentEnabled {
+		qrs.queue = jaegerqueue.NewBoundedQueue(qrs.cfg.QueueSize, func(item interface{}) {})
+	}
+	// The Persistent Queue is initialized separately as it needs extra information about the component
+
+	return qrs
+}
+
+func (qrs *queuedRetrySender) onSuccess(req request, err error) error {
+	return nil
+}
+
+func (qrs *queuedRetrySender) onPermanentFailure(req request, err error) error {
+	return err
+}
+
+func (qrs *queuedRetrySender) onTemporaryFailure(req request, err error) error {
+	if qrs.requeuingEnabled && qrs.queue != nil {
+		if qrs.queue.Produce(req) {
+			qrs.logger.Error(
+				"Exporting failed. Putting back to the end of the queue.",
+				zap.Error(err),
+			)
+		} else {
+			qrs.logger.Error(
+				"Exporting failed. Queue did not accept requeuing request. Dropping data.",
+				zap.Error(err),
+				zap.Int("dropped_items", req.count()),
+			)
+		}
+		return err
+	}
+
+	qrs.logger.Error(
+		"Exporting failed. No more retries left. Dropping data.",
+		zap.Error(err),
+		zap.Int("dropped_items", req.count()),
+	)
+	return err
```

Comment:
I'd personally invert the order here: happy path should be the least indented one (deal with errors first then perform the business logic you're meant to do)